overlord/servicestate/quota_control.go

```diff
@@ -473,5 +486,48 @@ func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) erro
 	}
 
 	// ensure service states are updated
-	return ensureSnapServicesForGroup(st, grp, allGrps)
+	return ensureSnapServicesForGroup(st, grp, allGrps, nil)
+}
+
+// EnsureSnapAbsentFromQuotaGroup ensures that the specified snap is not present
+// in any quota group, usually in preparation for removing that snap from the
+// system to keep the quota group itself consistent.
+func EnsureSnapAbsentFromQuotaGroup(st *state.State, snap string) error {
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	// try to find the snap in any group
+	var grp *quota.Group
+	for _, maybeGrp := range allGrps {
+		if strutil.ListContains(maybeGrp.Snaps, snap) {
+			grp = maybeGrp
+		}
+	}
+
+	if grp == nil {
+		return nil
+	}
+
+	// drop this snap from the snaps list of the group
+	newSnapList := make([]string, 0, len(grp.Snaps)-1)
+	for _, sn := range grp.Snaps {
+		if sn != snap {
+			newSnapList = append(newSnapList, sn)
+		}
+	}
+
+	grp.Snaps = newSnapList
+
+	// update the quota group state
+	allGrps, err = patchQuotas(st, grp)
+	if err != nil {
+		return err
+	}
+
+	// ensure service states are updated - note we have to add the snap as an
+	// extra snap to ensure since was removed from the group and thus won't be
```

Comment:
Did you mean "since *it* was removed"? Something reads weird here.

Reply:
yeah I rephrased this comment sorry it was confusing